File: IPC/message passing/java/ParentPipe.java
import java.io.*;

public class ParentPipe {

    public static void main(String[] args) throws Exception {

        ProcessBuilder pb = new ProcessBuilder("java", "ChildPipe");
        Process child = pb.start();

        BufferedReader childOut = new BufferedReader(
                new InputStreamReader(child.getInputStream()));
        BufferedWriter childIn = new BufferedWriter(
                new OutputStreamWriter(child.getOutputStream()));

        new Thread(() -> {
            try {
                String line;
                while ((line = childOut.readLine()) != null) {
                    System.out.println("Parent received: " + line);
                    childIn.write("Hello Child, message received by Parent\n");
                    childIn.flush();
                }
            } catch (IOException ignored) {}
        }).start();

        childIn.write("Hello from Parent\n");
        childIn.flush();

        child.waitFor();
        System.out.println("Child process exited");
    }
}

File: IPC/message passing/python/Parent_pipe.py
import subprocess
import threading

child = subprocess.Popen(
    ["python", "child_pipe.py"],
    stdin=subprocess.PIPE,
    stdout=subprocess.PIPE,
    text=True
)

def read_child():
    for line in child.stdout:
        print("Parent received:", line.strip())
        child.stdin.write("Hello Child, message received by Parent\n")
        child.stdin.flush()

threading.Thread(target=read_child, daemon=True).start()

child.stdin.write("Hello from Parent\n")
child.stdin.flush()

child.wait()
print("Child process exited")
